src/regenerate.rs

```diff
@@ -264,56 +265,100 @@ pub fn regenerate_doc_comments(directory: &str, verbose: bool) {
-            current_file = lines[it].replace(FILE, "").to_owned();
-            it += 1;
-        } else if lines[it].starts_with(MOD_COMMENT) {
-            let ty = parse_mod_line(&lines[it]);
-            let mut comments = vec!();
 
-            it += 1;
-            if ty.is_none() {
-                continue;
-            }
-            while it < lines.len() {
-                if lines[it].starts_with(MOD_COMMENT) ||
-                   lines[it].starts_with(FILE) {
-                    break;
+    let mut ret = HashMap::new();
+    let mut state = State::Initial;
+
+    while let Some(line) = lines.next() {
+        state = match state {
+            State::Initial => {
+                if line.starts_with(FILE) {
+                    State::File {
+                        file: line[FILE.len()..].to_owned(),
+                        infos: vec![],
+                        ty: None,
+                        comments: vec![],
+                    }
+                } else {
+                    panic!("Unrecognized format");
                 }
-                comments.push(lines[it].to_owned());
-                it += 1;
             }
-            if comments.len() > 0 {
-                current_infos.push((ty, comments));
-            }
-        } else if lines[it].starts_with(FILE_COMMENT) {
-            let mut comment_lines = vec!();
-
-            while it < lines.len() && lines[it].starts_with(FILE_COMMENT) {
-                comment_lines.push(lines[it][FILE_COMMENT.len()..].to_owned());
-                it += 1;
+            State::File { mut file, mut infos, mut ty, mut comments } => {
+                if line.starts_with(FILE) {
+                    if !comments.is_empty() {
+                        infos.push((ty.take(), comments));
+                        comments = vec![];
+                    }
+                    if !infos.is_empty() {
+                        ret.insert(file, infos);
+                        file = line[FILE.len()..].to_owned();
+                        infos = vec![];
+                    }
+                } else if line.starts_with(FILE_COMMENT) {
+                    if let Some(ty) = ty.take() {
+                        if !comments.is_empty() {
+                            infos.push((Some(ty), comments));
+                            comments = vec![];
+                        }
+                    } else {
+                        comments.push(line[FILE_COMMENT.len()..].to_owned());
+                    }
+                } else if line.starts_with(MOD_COMMENT) {
+                    if !comments.is_empty() {
+                        infos.push((ty, comments));
+                        comments = vec![];
+                    }
+                    ty = parse_mod_line(&line[..]);
+                } else {
+                    if ty.is_some() {
+                        comments.push(line[..].to_owned());
+                    } else {
+                        panic!("Orphan comment");
```

Comment:
I'm still not a big fan of panic. It doesn't really matter in here but I'd preferred that an error would be returned.